src/JsonReader.cs

```diff
@@ -548,127 +570,162 @@ sealed class NonProperty : IJsonProperty<Unit, JsonReadResult<Unit>>
+                            static bool TrySetPropertyIndex<TValue>(int index,
+                                                                    IJsonProperty<TValue, JsonReadResult<TValue>> property,
+                                                                    ref Utf8JsonReader reader,
+                                                                    ref ObjectValueState<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16> values)
+                            {
+                                if (!property.IsMatch(ref reader))
+                                    return false;
 
-                static bool ReadPropertyValue<TValue>(IJsonProperty<TValue, JsonReadResult<TValue>> property,
-                                                      ref Utf8JsonReader reader,
-                                                      ref (bool, TValue) value,
-                                                      ref string? error)
-                {
-                    if (!property.IsMatch(ref reader))
-                        return false;
+                                values.CurrentPropertyIndex = index;
+                                return true;
+                            }
 
-                    if (!reader.Read())
-                        throw PartialJsonNotSupportedException();
+                                 if (TrySetPropertyIndex(1,  property1,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(2,  property2,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(3,  property3,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(4,  property4,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(5,  property5,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(6,  property6,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(7,  property7,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(8,  property8,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(9,  property9,  ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(10, property10, ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(11, property11, ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(12, property12, ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(13, property13, ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(14, property14, ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(15, property15, ref reader, ref values)) { }
+                            else if (TrySetPropertyIndex(16, property16, ref reader, ref values)) { }
```

Comment:
nit: Potentially can be re-written as a short-circuiting expression to avoid the empty blocks (`{ }`):

```suggestion
                            _ =    TrySetPropertyIndex(1,  property1,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(2,  property2,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(3,  property3,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(4,  property4,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(5,  property5,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(6,  property6,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(7,  property7,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(8,  property8,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(9,  property9,  ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(10, property10, ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(11, property11, ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(12, property12, ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(13, property13, ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(14, property14, ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(15, property15, ref reader, ref values.CurrentPropertyIndex)
                                || TrySetPropertyIndex(16, property16, ref reader, ref values.CurrentPropertyIndex);
```